File: org.kbinani.vsq/VsqCommon.cs
/*
 * VsqCommon.cs
 * Copyright (C) 2008-2010 kbinani
 *
 * This file is part of org.kbinani.vsq.
 *
 * org.kbinani.vsq is free software; you can redistribute it and/or
 * modify it under the terms of the BSD License.
 *
 * org.kbinani.vsq is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
*/
#if JAVA
package org.kbinani.vsq;

import java.io.*;
import org.kbinani.*;
#else
using System;
using org.kbinani;
using org.kbinani.java.io;

namespace org.kbinani.vsq {

    using boolean = System.Boolean;
#endif

    /// <summary>
    /// vsqファイルのメタテキストの[Common]セクションに記録される内容を取り扱う
    /// </summary>
#if JAVA
    public class VsqCommon implements Cloneable, Serializable {
#else
    [Serializable]
    public class VsqCommon : ICloneable {
#endif
        public String Version;
        public String Name;
        public String Color;
        /// <summary>
        /// Dynamicsカーブを表示するモード(Expert)なら1、しない(Standard)なら0。
        /// </summary>
        public int DynamicsMode = org.kbinani.vsq.DynamicsMode.Expert;
        /// <summary>
        /// Play With Synthesisなら1、Play After Synthesiなら0、Offなら-1。
        /// </summary>
        public int PlayMode = org.kbinani.vsq.PlayMode.PlayWithSynth;
        /// <summary>
        /// PlayModeがOff(-1)にされる直前に，PlayAfterSynthかPlayWithSynthのどちらが指定されていたかを記憶しておく．
        /// </summary>
        public int LastPlayMode = org.kbinani.vsq.PlayMode.PlayWithSynth;

#if !JAVA
        public object Clone() {
            return clone();
        }
#endif

        public Object clone() {
            String[] spl = PortUtil.splitString( Color, new char[] { ',' }, 3 );
            int r = PortUtil.parseInt( spl[0] );
            int g = PortUtil.parseInt( spl[1] );
            int b = PortUtil.parseInt( spl[2] );
            VsqCommon res = new VsqCommon( Name, r, g, b, DynamicsMode, PlayMode );
            res.Version = Version;
            res.LastPlayMode = LastPlayMode;
            return res;
        }

        /// <summary>
        /// 各パラメータを指定したコンストラクタ
        /// </summary>
        /// <param name="name">トラック名</param>
        /// <param name="color">Color値（意味は不明）</param>
        /// <param name="dynamics_mode">DynamicsMode（デフォルトは1）</param>
        /// <param name="play_mode">PlayMode（デフォルトは1）</param>
        public VsqCommon( String name, int red, int green, int blue, int dynamics_mode, int play_mode ) {
            this.Version = "DSB301";
            this.Name = name;
            this.Color = red + "," + green + "," + blue;
            this.DynamicsMode = dynamics_mode;
            this.PlayMode = play_mode;
        }

#if JAVA
        public VsqCommon(){
            this(  "Miku", 179, 181, 123, 1, 1 );
#else
        public VsqCommon()
            : this( "Miku", 179, 181, 123, 1, 1 ) {
#endif
        }

        /// <summary>
        /// MetaTextのテキストファイルからのコンストラクタ
        /// </summary>
        /// <param name="sr">読み込むテキストファイル</param>
        /// <param name="last_line">読み込んだ最後の行が返される</param>
        public VsqCommon( TextStream sr, ByRef<String> last_line ) {
            Version = "";
            Name = "";
            Color = "0,0,0";
            DynamicsMode = 0;
            PlayMode = 1;
            last_line.value = sr.readLine();
            String[] spl;
            while ( !last_line.value.StartsWith( "[" ) ) {
                spl = PortUtil.splitString( last_line.value, new char[] { '=' } );
                String search = spl[0];
                if ( search.Equals( "Version" ) ) {
                    this.Version = spl[1];
                } else if ( search.Equals( "Name" ) ) {
                    this.Name = spl[1];
                } else if ( search.Equals( "Color" ) ) {
                    this.Color = spl[1];
                } else if ( search.Equals( "DynamicsMode" ) ) {
                    this.DynamicsMode = PortUtil.parseInt( spl[1] );
                } else if ( search.Equals( "PlayMode" ) ) {
                    this.PlayMode = PortUtil.parseInt( spl[1] );
                }
                if ( !sr.ready() ) {
                    break;
                }
                last_line.value = sr.readLine();
            }
        }

        /// <summary>
        /// インスタンスの内容をテキストファイルに出力します
        /// </summary>
        /// <param name="sw">出力先</param>
        public void write( ITextWriter sw ) {
            sw.writeLine( "[Common]" );
            sw.writeLine( "Version=" + Version );
            sw.writeLine( "Name=" + Name );
            sw.writeLine( "Color=" + Color );
            sw.writeLine( "DynamicsMode=" + DynamicsMode );
            sw.writeLine( "PlayMode=" + PlayMode );
        }
    }

#if !JAVA
}
#endif

File: port_js/org/kbinani/vsq/VsqCommon.js
/*
 * VsqCommon.js
 * Copyright (C) 2008-2010 kbinani
 *
 * This file is part of org.kbinani.vsq.
 *
 * org.kbinani.vsq is free software; you can redistribute it and/or
 * modify it under the terms of the BSD License.
 *
 * org.kbinani.vsq is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
*/
if( org == undefined ) var org = {};
if( org.kbinani == undefined ) org.kbinani = {};
if( org.kbinani.vsq == undefined ) org.kbinani.vsq = {};
if( org.kbinani.vsq.VsqCommon == undefined ){

    /**
     * vsqファイルのメタテキストの[Common]セクションに記録される内容を取り扱う
     */

    /**
     * overload1
     * @param text_stream [TextStream]
     * @param last_line [ByRef<String>]
     *
     * overload2
     * @param name [String]
     * @param red [int]
     * @param green [int]
     * @param blue [int]
     * @param dynamics_mode [int]
     * @param play_mode [int]
     */
    org.kbinani.vsq.VsqCommon = function(){
        this.Version = "DSB301";
        this.Name = "Miku";
        this.Color = "179,181,123";
        /**
         * Dynamicsカーブを表示するモード(Expert)なら1、しない(Standard)なら0。
         */
        this.DynamicsMode = org.kbinani.vsq.DynamicsMode.Expert;
        /**
         * Play With Synthesisなら1、Play After Synthesiなら0、Offなら-1。
         */
        this.PlayMode = org.kbinani.vsq.PlayMode.PlayWithSynth;
        /**
         * PlayModeがOff(-1)にされる直前に，PlayAfterSynthかPlayWithSynthのどちらが指定されていたかを記憶しておく．
         */
        this.LastPlayMode = org.kbinani.vsq.PlayMode.PlayWithSynth;
        if( arguments.length == 2 ){
            this._init_2( arguments[0], arguments[1], arguments[2], arguments[3] );
        }else if( arguments.length == 6 ){
            this._init_6( arguments[0], arguments[1], arguments[2], arguments[3], arguments[4], arguments[5] );
        }
    };

    org.kbinani.vsq.VsqCommon.prototype = {
        /**
         * @param sr [TextStream]
         * @param last_line [ByRef<string>]
         * @return [void]
         */
        _init_2 : function( sr, last_line ){
            this.Version = "";
            this.Name = "";
            this.Color = "0,0,0";
            this.DynamicsMode = 0;
            this.PlayMode = 1;
            last_line.value = sr.readLine();
            while ( last_line.value.charAt( 0 ) != "[" ) {
                var spl = last_line.value.split( "=" );
                var search = spl[0];
                if ( search == "Version" ) {
                    this.Version = spl[1];
                } else if ( search == "Name" ) {
                    this.Name = spl[1];
                } else if ( search == "Color" ) {
                    this.Color = spl[1];
                } else if ( search == "DynamicsMode" ) {
                    this.DynamicsMode = parseInt( spl[1], 10 );
                } else if ( search == "PlayMode" ) {
                    this.PlayMode = parseInt( spl[1], 10 );
                }
                if ( !sr.ready() ) {
                    break;
                }
                last_line.value = sr.readLine();
            }
        },

        _init_6 : function( name, r, g, b, dynamics_mode, play_mode ){
            this.Version = "DSB301";
            this.Name = arguments[0];
            this.Color = arguments[1] + "," + arguments[2] + "," + arguments[3];
            this.DynamicsMode = arguments[4];
            this.PlayMode = arguments[5];
        },

        clone : function() {
            var spl = this.Color.split( "," );
            var r = parseInt( spl[0], 10 );
            var g = parseInt( spl[1], 10 );
            var b = parseInt( spl[2], 10 );
            var res = new VsqCommon( this.Name, r, g, b, this.DynamicsMode, this.PlayMode );
            res.Version = this.Version;
            res.LastPlayMode = this.LastPlayMode;
            return res;
        },

        /**
         * インスタンスの内容をテキストファイルに出力します
         * @param sw [ITextWriter] 出力先
         * @return [void]
         */
        write : function( sw ){
            sw.writeLine( "[Common]" );
            sw.writeLine( "Version=" + this.Version );
            sw.writeLine( "Name=" + this.Name );
            sw.writeLine( "Color=" + this.Color );
            sw.writeLine( "DynamicsMode=" + this.DynamicsMode );
            sw.writeLine( "PlayMode=" + this.PlayMode );
        },
    };

}
